flameshot-org/flameshot
# Reduce dependence on D-Bus

Closes #1292.
Closes #1480.
Closes #1579.
Closes #1716.
As a side effect: Closes #1771.
May fix #1872.
May fix #1673.
My gut instinct says that this could be involved with #1246 somehow.

**NOTE: I tested nothing on a Mac**
### Terminology
- Daemon - The instance of flameshot that runs in the background
- Daemon hosting - The act of running the daemon in order to support some persistent phenomenon (pinned screenshot, clipboard, etc.)

### Things you should know:
- Flameshot does not support CLI arguments on Windows (and possibly Mac, can't test :cry:)
- When you create a pin, the daemon will host it. If it is not running
- The clipboard gets lost once flameshot quits on X11. Because of this, when any subcommand copies something to the clipboard, the daemon will be asked to host it. Will have to change this behavior for Windows and Mac.
- If there is no need to host something, the daemon will quit automatically.
  I.e. when all pinned widgets are closed, or when the clipboard is changed by another application.

### Changes and TODO (there will be more):
- [X] Pin and clipboard are hosted by the daemon
- [X] `flameshot screen` does not rely on dbus (except for hosting clipboard and pins)
- [X] `flameshot full` does not rely on dbus (...)
- [X] `flameshot gui` does not rely on dbus (...)
- [X] Do not show tray icon in non-dbus commands
- [X] Make `flameshot launcher` independent from dbus
- [X] Make `flameshot config` independent from dbus
- [x] `flameshot launcher` crashes when launching GUI
- [ ] ~Look into addressing #1672 in this~
- [X] "Copy file path after save" does not work properly
- [X] `flameshot gui` mutual exclusivity
- [X] Add option `autoCloseIdleDaemon`, unix-only, default: false
- [X] Tray icon should behave as it did before
- [X] Add the new config options to the GUI
- [ ] Edit test script for MacOS (Jeremy) (https://github.com/flameshot-org/flameshot/pull/2003#issuecomment-951047351)

### TODO (implementation details):
- [x] Reconcile `Controller` and `FlameshotDaemon` - make a clear distinction between them
- [x] Remove unnecessary arguments in Controller::captureTaken(Failed)
- [x] Clean up obsolete code

### TODO (when merging)
- [x] Keep track of notification-clipboard order as in #2053
- [X] Change --print-geometry format after #2013 is merged

### Testing:
- Before testing this PR, make sure the previous daemon version is killed
- **Everything should work as it did before this PR**
- If you have a Windows and/or Mac machine, please test thoroughly. I don't have
  a Mac machine.
- Run the scripts in the `tests` directory inside the repository root. These will test most of the CLI features, but not all!
- Abuse the `screen`, `full` and `gui` subcommands with various options and verify that the daemon is not launched, except when it needs to host the clipboard and/or pinned screenshots. You can use the script `tests/action_options.sh`
- Check if `flameshot gui` works correctly with various options and various types of manual interaction
- Test if stuff is copied to the clipboard correctly, e.g. paste into GitHub/gimp/etc and/or  run `xclip -o -t TARGETS -selection clip`
- If the daemon is not needed, it should quit automatically (on Linux). That is when all pinned screenshot windows are closed and the clipboard gets changed externally. On Windows and Mac, the daemon should keep running.
- Please test the heck out of daemon auto-launching, because this never worked
  on my system for some reason. In other words: make sure to **kill** the daemon before doing the other tests.

### To be discussed

- ~Do we allow multiple instances of `flameshot gui` to run? A benefit of this is the ability to take screenshots of flameshot itself.~
- ~Add a config option to keep daemon running even if not needed??~
- ~What should we do with the system tray if the daemon is running merely to host pins, clipboard, etc?~
